### PR TITLE
Implement GQA: Reduce Wallclock by 6% (WIP per YouJiacheng's review)

### PR DESCRIPTION
**Pull request by lapp0**
### Overview

This PR implements GQA, which reduces KV projection size and attention overhead. GQA decreases sample efficiency, requiring ~8% more training steps, but improving per-step performance by 14%.

Performance (`master` vs this PR):
- **runtime:** 447.86sec -> 422.81sec
- **Memory:** 34,237MiB -> 30,111MiB
- **Active Parameters / Token:** 123,573,558 -> 116,493,366
- **Total Parameters:** 394,002,486 -> 271,024,182

Large total parameter reduction primarily due to value embeddings halving in size, secondarily due to half the linear layer weights for the KV projection.

#### Baseline run logs:
- [step:1480/1480 val_loss:3.2818 train_time:448200ms step_avg:304.90ms](https://gist.github.com/lapp0/7aa8e2527a281bc12bbee663ff5a4d98)
- [step:1480/1480 val_loss:3.2768 train_time:447521ms step_avg:304.44ms](https://gist.github.com/lapp0/38b512f0b3f464c0e3c2eb6710a93a22)

#### GQA logs
- [step:1600/1600 val_loss:3.2790 train_time:423069ms step_avg:266.08ms](https://gist.github.com/lapp0/fc6d5e93a84a36608da996724299d3cc)
- [step:1600/1600 val_loss:3.2799 train_time:422214ms step_avg:265.54ms](https://gist.github.com/lapp0/4058a34b629413af1a74ca7df452bf10)
- [step:1600/1600 val_loss:3.2771 train_time:423154ms step_avg:266.13ms](https://gist.github.com/lapp0/7edb634b324cf81d52f97b3d338a55f3)

#### Hardware Advice?
I cannot reproduce the baseline on any 8xH100 NVL machine I've tried. In the record holders logs, the `step_avg` is about 1/2 my runs `step_avg`, Therefore % performance gain is relative to my own runs. I'd appreciate if someone could point me to a machine I could rent which could be used to reproduce the recordholders log.

### CODE REVIEW

**Comment by YouJiacheng** [comment]
image
H100 NVL has lower power wall and lower TFLOPS.
You can use NVIDIA H100 SXM 80GB.

**Comment by YouJiacheng** [comment]
Compilation time is sooo long that I haven't reproduced it yet.
But it seems that the most of performance boost comes from smaller value embedding.

**Comment by YouJiacheng** [comment]
My 3 runs gives 3.2851, 3.2811, 3.2755 val loss at ~233.5s (baseline is ~235.5s after some tweaks)
below are diffs compared to `master` (along with the diffs that make it ~235.5s)
```diff
@@ -172,9 +172,11 @@ class CausalSelfAttention(nn.Module):
         super().__init__()
         assert dim % num_heads == 0
         self.num_heads = num_heads
+        self.num_kv_heads = 3
+        kv_dim = self.num_kv_heads * (dim // num_heads)
         self.c_q = CastedLinear(dim, dim)
-        self.c_k = CastedLinear(dim, dim)
-        self.c_v = CastedLinear(dim, dim)
+        self.c_k = CastedLinear(dim, kv_dim)
+        self.c_v = CastedLinear(dim, kv_dim)
         self.lambdas = nn.Parameter(torch.tensor([0.5, 0.5]))
         self.rotary = Rotary(dim // num_heads) # dim // num_heads = head_dim
         self.c_proj = CastedLinear(dim, dim)
@@ -184,12 +186,12 @@ class CausalSelfAttention(nn.Module):
         B, T = x.size(0), x.size(1) # batch size, sequence length
         assert B == 1, "Must use batch size = 1 for FlexAttention"
         q = self.c_q(x).view(B, T, self.num_heads, -1)
-        k = self.c_k(x).view(B, T, self.num_heads, -1)
-        v = self.c_v(x).view(B, T, self.num_heads, -1)
+        k = self.c_k(x).view(B, T, self.num_kv_heads, -1)
+        v = self.c_v(x).view(B, T, self.num_kv_heads, -1)
         v = self.lambdas[0] * v + self.lambdas[1] * vi.view_as(v) # @KoszarskyB & @Grad62304977
         q, k = norm(q), norm(k) # QK norm @Grad62304977
         q, k = self.rotary(q), self.rotary(k)
-        y = flex_attention(q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), block_mask=block_mask)
+        y = flex_attention(q.transpose(1, 2), k.transpose(1, 2), v.transpose(1, 2), block_mask=block_mask, enable_gqa=True)
         y = y.transpose(1, 2).contiguous().view_as(x) # re-assemble all head outputs side by side
         y = self.c_proj(y)
         return y
@@ -234,7 +236,7 @@ class GPTConfig:
 
 class GPT(nn.Module):
 
-    def __init__(self, config):
+    def __init__(self, config: GPTConfig):
         super().__init__()
         self.num_layers = config.num_layers
 
@@ -248,11 +250,14 @@ class GPT(nn.Module):
         self.blocks = nn.ModuleList([Block(config) for _ in range(config.num_layers)])
         # token value embeddings by @KoszarskyB - inspired by @Grad62304977's value residual learning
         # U-net structure on token value embeddings by @leloykun
-        self.value_embeds = nn.Embedding(config.vocab_size, config.model_dim*self.num_encoder_layers)
+        self.value_embeds = nn.ModuleList([
+            nn.Embedding(config.vocab_size, config.model_dim // 2)
+            for _ in range(self.num_encoder_layers)
+        ])
         self.lm_head = CastedLinear(config.model_dim, config.vocab_size)
         self.lm_head.weight.data.zero_() # @Grad62304977
 
-    def forward(self, inputs, targets, sliding_window_size):
+    def forward(self, inputs: torch.Tensor, targets: torch.Tensor, sliding_window_size: torch.Tensor):
 
         BLOCK_SIZE = 128
         assert inputs.ndim == 1
@@ -285,7 +290,7 @@ class GPT(nn.Module):
         x = self.embed(inputs[None]) # token embeddings of shape (b, t, model_dim)
         x = norm(x) # @Grad62304977
         x0 = x
-        vi = self.value_embeds(inputs[None]).chunk(self.num_encoder_layers, dim=-1)
+        vi = [emb(inputs[None]) for emb in self.value_embeds]
 
         # Store outputs for U-Net skip connections
         skip_connections = []
@@ -375,7 +380,7 @@ class Hyperparameters:
     # optimization hyperparams
     batch_size : int = 8 # batch size, in sequences, across all devices
     sequence_length : int = 64*1024 # sequence length, in tokens
-    num_iterations : int = 1480 # number of iterations to run
+    num_iterations : int = 1600 # number of iterations to run
     warmup_iters : int = 0
     cooldown_iters : int = 600 # number of iterations of linear warmup/cooldown for triangular or trapezoidal schedule
     weight_decay : float = 0
@@ -401,7 +406,7 @@ logfile = None
 if master_process:
     run_id = str(uuid.uuid4())
     logdir = 'logs/%s/' % run_id
-    os.makedirs(logdir, exist_ok=True)
+    # os.makedirs(logdir, exist_ok=True)
     logfile = 'logs/%s.txt' % run_id
     # create the log file
     with open(logfile, "w") as f:
@@ -450,11 +455,12 @@ if hasattr(config, "coordinate_descent_tuning"):
     config.coordinate_descent_tuning = True # suggested by @Chillee
 model = torch.compile(model)
 # here we wrap model into DDP container
-model = DDP(model, device_ids=[ddp_local_rank])
+model = DDP(model, device_ids=[ddp_local_rank], gradient_as_bucket_view=True)
 raw_model = model.module # always contains the "raw" unwrapped model
 
 # init the optimizer(s)
-optimizer1 = torch.optim.Adam([raw_model.embed.weight, raw_model.value_embeds.weight], lr=0.6, betas=(0.8, 0.95), fused=True)
+embed_params = [*raw_model.embed.parameters(), *raw_model.value_embeds.parameters()]
+optimizer1 = torch.optim.Adam(embed_params, lr=0.6, betas=(0.8, 0.95), fused=True)
 optimizer2 = torch.optim.Adam([raw_model.lm_head.weight], lr=0.008, betas=(0.8, 0.95), fused=True)
 params = list(raw_model.blocks.parameters())
 matrix_params = [p for p in params if p.ndim == 2]
@@ -529,7 +535,7 @@ for step in range(args.num_iterations + 1):
         training_time_ms += 1000 * (time.perf_counter() - t0)
         # save the state of the training process
         log = dict(step=step, code=code, model=raw_model.state_dict(), optimizers=[opt.state_dict() for opt in optimizers])
-        torch.save(log, 'logs/%s/state_step%06d.pt' % (run_id, step))
+        # torch.save(log, 'logs/%s/state_step%06d.pt' % (run_id, step))
         # start the clock again
         torch.cuda.synchronize()
         t0 = time.perf_counter()
```

**Comment by lapp0** [comment]
Thanks so much for the hardware advice and cleanup!

Disappointing that the gain is marginal on SXM and your runs had much higher variance than mine. Will think about this for a bit.

**Comment by YouJiacheng** [comment]
Run additional 10 runs
image

**Comment by YouJiacheng** [comment]
btw I have tried shared value embedding across heads for mid-layers, without more training steps I can get ~3.285 loss but only ~6s faster.

**Comment by lapp0** [comment]
>val loss at ~233.5s (baseline is ~235.5s after some tweaks)

To clarify, you're only seeing a 2 second performance improvement with GQA enabled, all layers having 3 KV heads and 6 Q heads, all else the same?

**Comment by lapp0** [comment]
Looks like disabling GQA (setting `num_kv_heads=6`) benefits performance more for some layers than others.

image


Code:

<details>

```
from train_gpt2 import train, Hyperparameters, GPTConfig, master_process
from dataclasses import replace
import torch, gc


args = Hyperparameters(sequence_length=32768, batch_size=16)  # 8x4090
default_gpt_config = GPTConfig(flex_kernel_consumer=True)


# trials for full KV on these pairs
layer_pairs = [(0, 11), (1, 10), (2, 9), (3, 8), (4, 7), (5, 6), "all GQA", "no GQA"] + [(0, 2, 9, 11), (0, 3, 8, 11)]


for layer_pair in layer_pairs:
    if isinstance(layer_pair, tuple):
        layer_overrides = {
            layer: {"num_kv_heads": 6}
            for layer in layer_pair
        }
        gpt_config = replace(default_gpt_config, layer_overrides=layer_overrides)
    elif layer_pair == "all GQA":
        gpt_config = default_gpt_config
    elif layer_pair == "no GQA":
        gpt_config = replace(default_gpt_config, num_kv_heads=6)
    print(gpt_config)
    result = train(gpt_config, args)
    if master_process:
        with open("gqa_log.txt", "a") as f:
            f.write(f"{layer_pair}: {result}\n")

    del result
    gc.collect()
    torch.cuda.empty_cache()


# torchrun --standalone --nproc_per_node=8 gqa_trials.py
```

</details>

Will run a similar test to see how `num_intermediate` should be set by layer.

**Comment by lapp0** [comment]
A bit noisy, provides weak evidence that the last 3 layers benefit more than others from a 50% larger MLP `intermediate_dim`.

image

This comparison, which includes `num_kv_heads=6` for layers 0,11, further evidences that the `intermediate_dim` should be larger for the last 3 layers

image

Code and raw data

<details>

Chart 1:

```
from train_gpt2 import train, Hyperparameters, GPTConfig, master_process, ddp_rank
from dataclasses import replace
import torch, gc


args = Hyperparameters(sequence_length=32768, batch_size=16)  # 8x4090
default_gpt_config = GPTConfig(flex_kernel_consumer=True)


# trials for larger MLP on each layer
large_intermediate_dim_override = {"intermediate_dim": 768 * 6}


for layer_idx in range(12):
    gpt_config = replace(default_gpt_config, layer_overrides={layer_idx: large_intermediate_dim_override})
    print(gpt_config)

    result = train(gpt_config, args)

    if master_process:
        with open("mlp_log.txt", "a") as f:
            f.write(f"layer={layer_idx}: {result}\n")

    # cleanup
    del result
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.ipc_collect()
    torch.cuda.synchronize()
    torch._dynamo.reset()
```

```
data = {
    baseline': [3.2976, 3.2960, 3.2947, 3.2944],
    0: [3.2913, 3.2910,  3.2931],
    1: [3.2897, 3.2930, 3.2882],
    2: [3.2936, 3.2903, 3.2960],
    3: [3.2946, 3.2894, 3.2906],
    4: [3.2892, 3.2926, 3.2937],
    5: [3.2884, 3.2902, 3.2941],
    6: [3.2952, 3.2924, 3.2887],
    7: [3.2930, 3.2934, 3.2919],
    8: [3.2884, 3.2922, 3.2955],
    9: [3.2921, 3.2895, 3.2902],
    10: [3.2899, 3.2917, 3.2905],
    11: [3.2901, 3.2924, 3.2892],
}
```

Chart 2

```
from train_gpt2 import train, Hyperparameters, GPTConfig, master_process, ddp_rank
from dataclasses import replace
import torch, gc


args = Hyperparameters(sequence_length=32768, batch_size=16)  # 8x4090
default_gpt_config = GPTConfig(flex_kernel_consumer=True)


# trials for larger MLP on each layer
large_intermediate_dim_override = {"intermediate_dim": 768 * 6}


for layer_idx in range(12):
    gpt_config = replace(default_gpt_config, layer_overrides={layer_idx: large_intermediate_dim_override})
    print(gpt_config)

    result = train(gpt_config, args)

    if master_process:
        with open("mlp_log.txt", "a") as f:
            f.write(f"layer={layer_idx}: {result}\n")

    # cleanup
    del result
    gc.collect()
    torch.cuda.empty_cache()
    torch.cuda.ipc_collect()
    torch.cuda.synchronize()
    torch._dynamo.reset()
```

```
{"last 3 layers +1024": [3.2827, 3.2832, 3.2842], "all layers +256": [3.2858, 3.2840, 3.2854]}
```

</details>

**Comment by YouJiacheng** [comment]
FYI, I ran experiments that different layers has different "number of heads" of value embedding. the conclusion is similar: the mid-layers are less important.

**Comment by lapp0** [comment]
I see a 2% reduction in runtime with the configuration below, but it complicates the model quite a bit and GQA has absurd compile times. I'll think about this some more.

```
+    layer_overrides: Dict[int, Dict] = field(default_factory=lambda: {
+        0: {"num_kv_heads": 6},
+        2: {"num_kv_heads": 6},
+        9: {"num_kv_heads": 6, "intermediate_dim": 4096},
+        11: {"num_kv_heads": 6, "intermediate_dim": 4096},
+        10: {"intermediate_dim": 4096},
+    })
```
